### PR TITLE
[WIP]RSP-1565: ButtonGroup to use collection Item and context

### PR DESCRIPTION
**Pull request by pst67662**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [Issue](https://github.com/adobe-private/react-spectrum-v3/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-stately/collections/src/CollectionBuilder.tsx [hunk 1]
```diff
@@ -10,12 +10,19 @@
...
-  return null;
+// export function Item<T>(props: ItemProps<T>): ReactElement { // eslint-disable-line @typescript-eslint/no-unused-vars
+//   return null;
+// }
+
+export const CollectionsContext = React.createContext<CollectionsContextValue | null>(null);
+
+export function Item<T>(props: ItemProps<T>): ReactElement {
+  let {itemComponent: ItemComponent} = useContext(CollectionsContext);
+  return <ItemComponent {...props} />;
```
Comment (pst67662) [inline packages/@react-stately/collections/src/CollectionBuilder.tsx]: ticket description says:
```
function Item({children}) {
  let {itemComponent: ItemComponent} = useContext(CollectionsContext);
  return <ItemComponent>{children}</ItemComponent>;
}
```
but I think that other props will be lost with just `<ItemComponent>{children}</ItemComponent>` ?

File: packages/@react-types/shared/src/collections.d.ts [hunk 2]
```diff
@@ -141,3 +141,8 @@ interface AsyncListProps<T> {
 }
 
 declare function useAsyncList<T>(opts: AsyncListOptions<T>): AsyncListProps<T>;
+
+export interface CollectionsContextValue {
+  sectionComponent?: string | JSXElementConstructor<any>,
+  itemComponent?: string | JSXElementConstructor<any>
+}
```
Comment (pst67662) [inline packages/@react-types/shared/src/collections.d.ts]: should props be optional with default? something like
itemComponent: ItemComponent = 'div'

File: packages/@react-stately/button/src/useButtonGroupState.ts [hunk 3]
```diff
@@ -16,12 +16,13 @@ import {MultipleSelection} from '@react-types/shared';
 import React, {useMemo} from 'react';
 import {SelectionManager, useMultipleSelectionState} from '@react-stately/selection';
 
-export function useButtonGroupState(props: ButtonGroupStateBase & MultipleSelection): ButtonGroupState {
+export function useButtonGroupState<T>(props: ButtonGroupStateBase<T> & MultipleSelection): ButtonGroupState {
   let selectionState = useMultipleSelectionState(props);
 
   let buttonCollection = useMemo(() => {
     let childrenArray = React.Children.toArray(props.children);
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-stately/button/src/useButtonGroupState.ts]: I don't think that's necessary for ButtonGroup. It's relatively simple so doesn't need this complexity.

File: packages/@react-spectrum/button/src/ButtonGroup.tsx [hunk 4]
```diff
@@ -60,40 +61,35 @@ export function ButtonGroup<T>(props: CollectionBase<T> & SpectrumButtonGroupPro
...
-            <ButtonGroupItem
-              key={item.key}
-              {...buttonProps}
-              className={classNames(buttonStyles, 'spectrum-ButtonGroup-item')}
-              item={item}
-              state={state} />
-          ))
-        }
+        <CollectionsContext.Provider
+          value={{itemComponent: (item) => <ButtonGroupItem state={state} item={item} {...buttonProps} />}}>
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-spectrum/button/src/ButtonGroup.tsx]: Seems ok to me

File: packages/@react-spectrum/button/src/ButtonGroup.tsx [hunk 5]
```diff
@@ -60,40 +61,35 @@ export function ButtonGroup<T>(props: CollectionBase<T> & SpectrumButtonGroupPro
...
+          {state.buttonCollection.items}
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-spectrum/button/src/ButtonGroup.tsx]: Should be `children`. You could have wrappers around the actual items, for example `<DialogTrigger>`, so the items don't map 1:1 with children.

File: packages/@react-stately/collections/src/CollectionBuilder.tsx (hunk 1 above)
Comment (devongovett) [reply packages/@react-stately/collections/src/CollectionBuilder.tsx]: makes sense.

File: packages/@react-types/shared/src/collections.d.ts (hunk 2 above)
Comment (devongovett) [reply packages/@react-types/shared/src/collections.d.ts]: Eh... I think doing nothing by default is ok for now. When used inside a normal collection component we don't use the result of the rendering anyway.

File: packages/@react-spectrum/button/src/ButtonGroup.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (pst67662) [reply packages/@react-spectrum/button/src/ButtonGroup.tsx]: I didnt pass item correctly
` value={{itemComponent: (propsFromChild) => <ButtonGroupItem state={state} {...propsFromChild} />}}>`

so, I would need to pair child with correct item from collection?

**Comment by devongovett** [comment]
I think this PR was superseded by the collection builder one.

**Pull request closed without merging**